proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+
+#### 1. Contracts are untrustworthy by default
+```javascript
+// bad
+Bank.withdraw(100); // Unclear whether trusted or untrusted
+
+function makeWithdrawal(uint amount) { // Isn't clear that this function is potentially unsafe
+    Bank.withdraw(amount);
+}
+
+// good
+UntrustedBank.withdraw(100); // untrusted external call
+TrustedBank.withdraw(100); // external but trusted bank contract maintained by XYZ Corp
+
+function makeUntrustedWithdrawal(uint amount) {
+    UntrustedBank.withdraw(amount);
+}
+```
+
+#### 2. External calls have arbitrary code execution
+Examples of problems that can arise in Solidity code include:
+##### Race Conditions
+Whether using raw calls (of the form `someAddress.call()`) or contract calls (of the form `ExternalContract.someMethod()`), malicious code might execute. Even if ExternalContract is not malicious, malicious code can be executed by any contracts it calls.
+One particular danger is malicious code may hijack the control flow, leading to race conditions.
+
+##### Re-entrancy
+- `someAddress.send()` and `someAddress.transfer()` in Solidity are considered safe against reentrancy. While these methods still trigger code execution, the called contract is only given a stipend of 2300 gas which is currently only enough to log an event.
+  - Prevents reentrancy but is incompatible with any contract whose fallback function requires 2300 gas or more.
+  - Sometimes the programmer won't want this, but then has to fall back onto the dangerous raw calls.
+
+#### 3. External calls can silently fail
+Solidity offers low-level call methods that work on rawAddress: `address.call()`, `address.callcode()`, `address.delegatecall()`, `address.send()`. These low-level methods never throw an exception so they fail silently.
+
+```javascript
+// bad
+someAddress.send(55);
+// this is doubly dangerous, as it will forward all remaining gas and doesn't check for result
+someAddress.call.value(55)(); 
+// if deposit throws an exception, the raw call() will only return false and transaction will NOT be reverted
+someAddress.call.value(100)(bytes4(sha3("deposit()"))); 
+
+// good
+if(!someAddress.send(55)) {
+    // Some failure code
+}
+
+ExternalContract(someAddress).deposit.value(100);
+```
+
+#### 4. Interfaces can easily be incorrectly specified
+Minor errors in Solidity interfaces may lead to wrong code being executed. `Alice.set(uint)` takes an `uint` in `Bob.sol` but `Alice.set(int)` takes an `int` in `Alice.sol`. The two interfaces will produce two differents method IDs. As a result, Bob will call the fallback function of Alice rather than of `set`. This type of error is responsible for the King of the Ether bug.
```

Comment:
Is it also possible for an interface to match when it shouldn't in the case of a minor error - is type information lost in Yul translation for interfaces?